### PR TITLE
Transcripts: search logic

### PR DESCRIPTION
**Pull request by leandroalonso**
| 📘 Part of: #1848 | 
|:---:|

| iPhone | iPad |
| ------- | ---- |
| Simulator Screenshot - iPhone 15 - 2024-08-02 at 15 24 31 | Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-08-02 at 15 24 57 |

This adds the search logic inside a single transcript. I implemented a version of the [Knuth–Morris–Pratt algorithm](https://en.wikipedia.org/wiki/Knuth–Morris–Pratt_algorithm) to be as performant as possible.

## To test

Before testing, ensure **transcripts** feature flag is enabled. It's also better to test on a real device. If you want to test on the simulator, use the simulator keyboard, not your Mac's.

### Testing UI componentes

1. Run the app
2. Play any episode with transcripts, [there's a list here](https://lists.pocketcasts.com/b852a088-ccee-4e4b-a3c5-4bb7fd700a02)
3. Tap the Transcript button and then tap "Search"
4. Scroll down to the end of the transcript
5. ✅ The scroll view should end when the keyboard starts
6. Tap "Done"
7. ✅ The scroll view should get back to its original position
8. Test the same on iPad, but when the keyboard is appearing toggle between portrait/landscape and ensure the scroll view is adjusted accordingly

### Testing search

1. Run the app
2. Play any episode with transcript, [there's a list here](https://lists.pocketcasts.com/b852a088-ccee-4e4b-a3c5-4bb7fd700a02)
3. Tap the transcript button and then tap "Search"
4. Enter any word
5. ✅ The app scrolls to the first results of your search
6. ✅ The input accessory should display "1 of X" where X is the number of results
7. ✅ The first result should have a white highlight color and black text color
8. ✅ The rest of the results should have a white background color with opacity
9. Tap the chevron down or up to toggle between results
10. ✅ The app should scroll to the next result updating which one is highlighted
11. ✅ When you reach the last one and tap the chevron down, it should go to the first
12. ✅ When showing the first result and you tap the chevron up, it should go to the last
13. ✅ The label should update indicating which result you see, eg.: "3 of 10"
14. Tap "search" on the keyboard
15. ✅ The keyboard should disappear, but the results are still highlighter and the input accessory is at the bottom of the screen
16. ✅ Ensure you can still use the chevrons up and down
17. Tap "Done"
18. ✅ All highlights are removed
19. Tap "Search" again and search for something
20. Tap "Done"
21. ✅ Keyboard, input accessory, and highlights should all disappear

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UITextView+scrollToRange.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+extension UITextView {
+    public func scrollToRange(_ range: NSRange) {
+        // Ensure layout is up-to-date
+        textStorage.addLayoutManager(layoutManager)
+        layoutManager.ensureLayout(for: textContainer)
+
+        let glyphRange = layoutManager.glyphRange(forCharacterRange: range, actualCharacterRange: nil)
+
+        layoutManager.enumerateEnclosingRects(forGlyphRange: glyphRange, withinSelectedGlyphRange: NSRange(location: NSNotFound, length: 0), in: textContainer) { rect, _ in
```
Comment (danielebogo) [inline Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UITextView+scrollToRange.swift]: Just double check if a retain cycle happens here as the `enumerateEnclosingRects` closure is `escaping`

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#if !os(watchOS)
+import UIKit
+
+extension UIView {
+    public func distanceFromBottom() -> CGFloat? {
+        guard let window = self.window else {
```
Comment (danielebogo) [inline Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift]: Nit: You can simplify it as
```
guard let window else {
```

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+            return nil
+        }
+
+        let viewFrameInWindow = self.convert(self.bounds, to: window)
```
Comment (danielebogo) [inline Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift]: Nit: You can avoid `self.` in both places

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+        let screenHeight = UIScreen.main.bounds.height
+        let distanceFromBottom = screenHeight - viewFrameInWindow.maxY
```
Comment (danielebogo) [inline Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift]: Nit: I guess you can just `return screenHeight - viewFrameInWindow.maxY` directly if you don't need `distanceFromBottom`

File: podcasts/TranscriptsViewController.swift [hunk 5]
```diff
@@ -361,6 +402,77 @@ class TranscriptsViewController: PlayerItemViewController {
...
+        }
+
+        let kmpSearch = KMPSearch(pattern: term)
+        searchIndicesResult = kmpSearch.search(in: transcriptText)
+        searchTerm = term
+    }
+
+    @MainActor
+    func updateNumberOfResults() {
+        guard searchIndicesResult.count > 0 else {
```
Comment (danielebogo) [inline podcasts/TranscriptsViewController.swift]: Same here about the `guard`

File: podcasts/TranscriptsViewController.swift [hunk 6]
```diff
@@ -390,12 +502,48 @@ extension TranscriptsViewController: UIScrollViewDelegate {
...
+        resetSearch()
         searchView.removeFromSuperview()
     }
 
     func searchButtonTapped() {
         becomeFirstResponder()
     }
+
+    func search(_ term: String) {
+        guard !term.isEmpty else {
```
Comment (danielebogo) [inline podcasts/TranscriptsViewController.swift]: As much as I love `guard` I would avoid to use it as double negative. In these cases `if` is more readable, as `if term.isEmpty`

File: Modules/Utils/Sources/PocketCastsUtils/General/KPMSearch.swift [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    private var lps: [Int]
+
+    public init(pattern: String) {
+        self.pattern = Array(pattern.lowercased())
+        self.lps = Array(repeating: 0, count: pattern.count)
+        computeLPSArray()
+    }
+
+    private func computeLPSArray() {
+        guard lps.count > 0 else {
```
Comment (danielebogo) [inline Modules/Utils/Sources/PocketCastsUtils/General/KPMSearch.swift]: I would use `if lps.isEmpty` directly.
In general with collections, `isEmpty` is more efficient than `count` as it doesn't loop trough the whole list.

File: Modules/Utils/Sources/PocketCastsUtils/General/KPMSearch.swift [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                } else {
+                    lps[i] = 0
+                    i += 1
+                }
+            }
+        }
+    }
+
+    public func search(in text: String) -> [Int] {
+        guard pattern.count > 0 else {
```
Comment (danielebogo) [inline Modules/Utils/Sources/PocketCastsUtils/General/KPMSearch.swift]: Same here about the `guard`

**Review by danielebogo — CHANGES_REQUESTED** [review]
Hey @leandroalonso this looks really great!
I left some comments, most are NIT.

During the test thought I encountered a crash:
• Podcast: Cautionary Tales - Embracing the Escape Fire
• I opened the transcript and searched for `They`
• I used the chevron to navigate to the 30/40th element found
• Then I selected the word in the search bar and started typing `defining`
• The search shows 36 of 1 as result and as soon I tap the chevron the app crashes

Screenshot 2024-08-05 at 10 29 49
Screenshot 2024-08-05 at 10 36 37

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UITextView+scrollToRange.swift (hunk 1 above)
Comment (leandroalonso) [reply Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UITextView+scrollToRange.swift]: Good catch! Fixed in 1999d3f868649eff31832c3dbf2805ef918a799a

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift (hunk 2 above)
Comment (leandroalonso) [reply Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift]: Done in c237ddff541659d483e2ae3837afedf8ad7ef450

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift (hunk 3 above)
Comment (leandroalonso) [reply Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift]: Done in c237ddff541659d483e2ae3837afedf8ad7ef450

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift (hunk 4 above)
Comment (leandroalonso) [reply Modules/Utils/Sources/PocketCastsUtils/Extensions/UI/UIView+distanceToBottom.swift]: Done in c237ddff541659d483e2ae3837afedf8ad7ef450

File: podcasts/TranscriptsViewController.swift (hunk 5 above)
Comment (leandroalonso) [reply podcasts/TranscriptsViewController.swift]: Done in b351ceeb690ee61d6f813f00e0b91012daa36bff

File: podcasts/TranscriptsViewController.swift (hunk 6 above)
Comment (leandroalonso) [reply podcasts/TranscriptsViewController.swift]: Done in ee793f555a4e572d99d31f421cb5585a1b34e94d

File: Modules/Utils/Sources/PocketCastsUtils/General/KPMSearch.swift (hunk 7 above)
Comment (leandroalonso) [reply Modules/Utils/Sources/PocketCastsUtils/General/KPMSearch.swift]: Done in 4fee0685bdb7e08cb29169c2c6e15d3b42b0732f

File: Modules/Utils/Sources/PocketCastsUtils/General/KPMSearch.swift (hunk 8 above)
Comment (leandroalonso) [reply Modules/Utils/Sources/PocketCastsUtils/General/KPMSearch.swift]: Done in 1dcaa25016d61417b07fd8165d3fab0a4434ede5

**Comment by leandroalonso** [comment]
@danielebogo I wasn't able to reproduce the crash. However, there was a logic error as the current index should update when the search is updated. Could you please review it again? Thanks!

**Review by danielebogo — APPROVED** [review]
Awesome @leandroalonso !
I can't reproduce the issue anymore! Thanks a lot for addressing all the nit! :shipit:

**Comment by leandroalonso** [comment]
@pachlava not sure what are your priorities right now but I'd love some testing on the search functionality once you have a chance. Thank you!

**Comment by pachlava** [comment]
Thanks for the ping @leandroalonso! I'll have a look into search tomorrow.

**Comment by pachlava** [comment]
@leandroalonso Not sure if it's the best place to communicate an issue, since the PR is merged. 

If I enter search mode, and then dismiss the keyboard by tapping somewhere on the transcript (no matter if I entered search term or not), the scroll position would jump seemingly unpredictably:

https://github.com/user-attachments/assets/716cba21-a5ce-4620-87d4-d791c1597e35

**Pull request merged**